src/contract.rs

```diff
@@ -298,6 +294,22 @@ fn approve_ask(
                     fields: vec![String::from("id")],
                 }),
                 Some(mut stored_ask_order) => {
+                    // Validate ask order hasnt been approved yet
+                    match stored_ask_order.class {
+                        AskOrderClass::Convertible { status } => match status {
+                            AskOrderStatus::Ready {
```

Comment:
is this the only status we need to check against or can it be in another status?

Reply:
only can be two statuses, `Ready` or `PendingIssuerApproval`

Reply:
i could have the match check excatly those two instead of `_` to throw a compile error if more are added later

Reply:
this may be a newbie (to smart contracts) question but... if we need to check for the status do we also need to worry about race conditions (is the contract locked in a transaction while this code is executing)?

Reply:
i'm pretty sure the way the smart contracts work is that transactions gets processed one by one in sequential order that they are submit to the blockchain.
if they fail, the transaction fails and the requester would have to re-submit it. I dont think there would be a contract "locking" in the traditional sense like a database

if there is some sort of a race condition, its would kind of be like:
1. Transaction 1 comes in to do X
2. Transaction 2 comes in to do X
3. A provenance validator attempts transaction 1 and then after completion, tries to do transaction 2

I think this is a detailed doc of the transaction lifecycle, https://docs.provenance.io/blockchain/basics/transaction-lifecycle

This is how i would explain a transaction composition
- Msgs make up a Tx, they are ordered by the requester
- Txs make up a Block (i think they are consumed sequentially, in the order they were receive into the blockchain Mempool)
- Blocks are sequential and make up the whole blockchain

Reply:
Thank you for the thorough information and links! I really appreciate the knowledge sharing!